app/src/main/kotlin/compiler/intermediate_form/FunctionDetailsGenerator.kt

```diff
@@ -1,33 +1,250 @@
+    val variableToStoreFunctionResult: Variable?,
+    val functionLocationInCode: IntermediateFormTreeNode.MemoryAddress,
     val depth: ULong,
-    val vars: Map<NamedNode, Boolean>,
-    val parameters: List<Function.Parameter>
+    val variablesLocationTypes: Map<NamedNode, VariableLocationType>, // should contain parameters
+    val displayAddress: IntermediateFormTreeNode,
+    val createRegisterFor: (NamedNode) -> Register = { Register() } // for testing
 ) : FunctionDetailsGeneratorInterface {
 
-    override fun generateCall(args: List<IntermediateFormTreeNode>): FunctionDetailsGeneratorInterface.FunctionCallIntermediateForm {
-        return TODO()
+    private val variablesStackOffsets: MutableMap<NamedNode, ULong> = ReferenceHashMap()
+    private val variablesRegisters: MutableMap<NamedNode, Register> = ReferenceHashMap()
+    private var variablesTotalOffset: ULong = 0u
+    private val previousDisplayEntryRegister = Register()
+
+    init {
+        for ((variable, locationType) in variablesLocationTypes.entries) {
+            when (locationType) {
+                VariableLocationType.MEMORY -> {
+                    variablesStackOffsets[variable] = variablesTotalOffset
+                    variablesTotalOffset += memoryUnitSize
+                }
+
+                VariableLocationType.REGISTER -> {
+                    variablesRegisters[variable] = createRegisterFor(variable)
+                }
+            }
+        }
+    }
+
+    override fun genCall(
+        args: List<IntermediateFormTreeNode>,
+    ): FunctionDetailsGeneratorInterface.FunctionCallIntermediateForm {
+        // First, it moves arguments to appropriate registers (or pushes to stack) according to x86 call convention.
+        // Then, it adds call instruction to actually call a given function
+        // At the end it creates IFTNode to get function result
+
+        val cfgBuilder = ControlFlowGraphBuilder()
+
+        for ((arg, register) in args zip argPositionToRegister) {
+            val node = IntermediateFormTreeNode.RegisterWrite(register, arg)
+            cfgBuilder.addLinkFromAllFinalRoots(CFGLinkType.UNCONDITIONAL, node)
+        }
+        for (arg in args.drop(argPositionToRegister.size).reversed()) {
+            val node = IntermediateFormTreeNode.StackPush(arg)
+            cfgBuilder.addLinkFromAllFinalRoots(CFGLinkType.UNCONDITIONAL, node)
+        }
+
+        cfgBuilder.addLinkFromAllFinalRoots(
+            CFGLinkType.UNCONDITIONAL,
+            IntermediateFormTreeNode.Call(functionLocationInCode)
+        )
+
+        var readResultNode: IFTNode? = null
+        if (variableToStoreFunctionResult != null) {
+            cfgBuilder.addLinkFromAllFinalRoots(
+                CFGLinkType.UNCONDITIONAL,
+                IntermediateFormTreeNode.RegisterWrite(Register.RAX, genRead(variableToStoreFunctionResult, true))
+            )
+            readResultNode = IntermediateFormTreeNode.RegisterRead(Register.RAX)
```

Comment:
True, I've moved saving into rax to `genEpilogue`. 
Although, we need `variableToStoreFunctionResult` and creating such temporary variable in `ControlFlow.createGraphForEachFunction`, because we need to store somewhere the result and `createGraphForEachFunction` shouldn't know anything about `Register` backend, so we'd rather not use `RAX` name there. Alternatively, we could pass the `Register` to that function to which it should write the result, but introducing a new `Variable` is cleaner imho, as it moves the responsibility to decide the actual Register (or memory address) for such variable to graph coloring algorithm we'll be using for allocating physical registers.

Reply:
OK, I guess I didn't understand the function of that variable well enough. As for the `Variable` vs `Register`, using a `Register` would also let the coloring algorithm decide (as long as we pass a virtual register, not one of the physical ones). But since we already have means to write the expression result to a variable, I agree that there's no need to change that.
One thing that bothers me is that we create new `Variable` object after all the semantic analysis phases, consequently making some maps incorrect (incomplete). It seems that it's not a problem now, but it has a potential to become a problem in the future.

Reply:
Also, function can have multiple 'return' statements and we want to use the same variable in all of them, so we should create it only once.

Reply:
Moved variable creation to the beginning of this file to make sure it is done only once. For semantic analysis, we can create a separate issue for it, maybe linked with passing additional parameters to creating `DefaultFunctionDetailsGenerator` in `FunctionDependenciesAnalyzer.createFunctionDetailsGenerators`, because we have to pass that variable to the constructor of `DFDG` somehow anyway.